Detect if Passkey Challenge Was Completed on the Current Device

malwaregeeek opened the pull request:
## **What**

Adds the `authenticatorAttachment` field to the WebAuthn authentication response, allowing detection of whether a passkey challenge was completed using a platform authenticator (current device) or through cross-device authentication.

Closes #42931

##  **Why**

Currently, there is no reliable way to detect whether a user completes the passkey challenge on their current device or via a different device. This information is crucial for:

- **Conditional enrollment flows**: Show WebAuthn enrollment screen only when no passkeys are found on the current device
- **Improved UX**: Skip enrollment prompts when platform authenticators are available
- **Multi-device support**: Better handling of passkey registration across multiple devices

## **How** 
The authenticatorAttachment field can return one of the following

**platform** → authenticators built into the device (e.g. Touch ID, Face ID, Windows Hello, Android fingerprint).

**cross-platform** → roaming authenticators (e.g. USB/Bluetooth/NFC security keys like YubiKey).

ahus1 commented:
Pinging the @keycloak/core-clients team as well as this is passkeys related.

rmartinc commented on themes/src/main/resources/theme/base/login/resources/js/webauthnAuthenticate.js:
```diff
@@ -91,6 +91,9 @@ export function returnSuccess(result) {
     if (result.response.userHandle) {
         document.getElementById("userHandle").value = base64url.stringify(new Uint8Array(result.response.userHandle), { pad: false });
     }
+    if (result.response.platformAttachment) {
```
If I'm not wrong this should be [result.authenticatorAttachment](https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredential/authenticatorAttachment) which can have the values `platform` or `cross-platform`. Or am I missing something?

rmartinc reviewed (changes requested):
@malwaregeeek I think that the property is incorrect and it should be `result.authenticatorAttachment`. Right?

I also think that this PR does nothing for the moment. I don't know if you want to do something more in subsequent PRs but, with the current change, no new feature or enhancement is achieved. I would say that you at least need to recover the passed property at server side and attach it in the login event. This way you can create a test that checks that the `authenticatorAttachment` is at least shown in the event with the correct value (platform or cross-platform). If you want to achieve what is proposed in  https://github.com/keycloak/keycloak/issues/40988 you will need more PRs.

malwaregeeek committed:
Add authenticatorAttachment field support to WebAuthn authentication

Detects if passkey challenge was completed using the current device (platform authenticator) or other external authenticator

Closes https://github.com/keycloak/keycloak/issues/42931

Signed-off-by: ankuragrawal420 <ankuragrawal420@gmail.com>

malwaregeeek commented:
> @malwaregeeek I think that the property is incorrect and it should be `result.authenticatorAttachment`. Right?
> 
> I also think that this PR does nothing for the moment. I don't know if you want to do something more in subsequent PRs but, with the current change, no new feature or enhancement is achieved. I would say that you at least need to recover the passed property at server side and attach it in the login event. This way you can create a test that checks that the `authenticatorAttachment` is at least shown in the event with the correct value (platform or cross-platform). If you want to achieve what is proposed in #40988 you will need more PRs.

@rmartinc Thanks for reviewing the PR! You’re right — that was a typo, it should have been `authenticatorAttachment` (now corrected). This PR doesn’t change any behavior on its own; it’s more of a preparatory step. That said, it could still be useful for folks customizing the `WebAuthnAuthenticator`, since they wouldn’t need to modify `webauthnAuthenticate.js.`

My plan is to follow up with another PR that updates the `WebAuthnAuthenticator` to read this additional form field and set an `authNote` with the authenticatorAttachment type. That authNote can then be used in `WebAuthnRegisterRequiredAction` (in the `evaluateTrigger` method) to decide when the registration prompt should be shown.

pnzrr commented:
@rmartinc as @malwaregeeek described, it's as much preparatory as anything, the current webauthnAuthenticate.js isn't passing the whole object through, which if you want to use missing parts, means you have to fork the file and add to it, when it should just do that be default.

malwaregeeek committed:
Add test for authenticatorAttachment field during WebAuthn authentication

- Tests that authenticatorAttachment field is present in WebAuthn login form
- Verifies field exists for both platform and cross-platform authenticators
- Uses virtual authenticators with INTERNAL (platform) and USB (cross-platform) transports
- Ensures the form field integration works correctly with different authenticator types

Related to GitHub issue #42931

Signed-off-by: ankuragrawal420 <ankuragrawal420@gmail.com>

malwaregeeek committed:
Simplify WebAuthn authenticatorAttachment test

- Remove cross-platform authenticator test portion
- Focus test on platform authenticator only
- Verify authenticatorAttachment field exists and authentication succeeds
- Simplified cleanup and test flow

Signed-off-by: ankuragrawal420 <ankuragrawal420@gmail.com>

malwaregeeek committed:
Update WebAuthn test to read authenticatorAttachment field value

- Test now reads and validates the authenticatorAttachment field value
- Verifies field is set to 'platform' for platform authenticators
- Provides better validation of the authenticatorAttachment implementation

Signed-off-by: ankuragrawal420 <ankuragrawal420@gmail.com>

malwaregeeek committed:
Remove standalone authenticatorAttachment test

- Removed testAuthenticatorAttachmentFieldInAuthentication method
- Will integrate authenticatorAttachment validation into existing test instead

Signed-off-by: ankuragrawal420 <ankuragrawal420@gmail.com>
